File: src/lane_detection/src/ld_scanner.cpp
/**
 * @file ld_scanner.h
 * @brief This file contains the implementation file for Ld_scanner class for
 scanning for lane lines segments in a image.
 *
 * @author Trevin Vaughan
 * @date 2025-07-28
 *
 * @par License:
 *      MIT License
 *
 * @par Description:
 *      - This class takes an binary image and scans for lane line segments base
 *        on it width and position in the image.
 *      - Does a median filter on the scan results
 *
 * @par Requirements:
 *      - OpenCV >= 4.5.0
 *      - C++17 or higher
 *      - ROS2 Humble
 *
 * @par Changelog:
 *      - 2025-07-31: Initial creation
 */

#include "../include/lane_detection/ld_scanner.h"

#include <opencv2/core/mat.hpp>
#include <opencv2/core/types.hpp>
#include <rclcpp/logger.hpp>
#include <rclcpp/logging.hpp>

namespace ld_scanner_ns {
// ============== Adding Names ==============================================
// std naming
using std::max;
using std::min;
using std::pair;

// ============== Main Interface Function Implementations =====================
bool Ld_scanner::run(const Mat &bin_mask) {
    // reset the lane pixel coordinate vectors
    this->left_lane.clear();
    this->right_lane.clear();
    this->center_lane.clear();

    for (int y = 0; y < bin_mask.rows; y += config.steps) {
        vector<pair<int, int>> lane_segments;

        bool in_lane = false;    // Is the index inside a lane segment?
        int segment_start = -1;  // lane segment start x coordinate

        // Find all white pixel (lane markings) in this row
        for (int x = 0; x < bin_mask.cols; x++) {
            // opencv stores black and white image as black = 0 and white =
            // 255.
            uchar pixel_value = bin_mask.at<uchar>(y, x);
            if (lane_segments.size() == 2) {
                // only need need two lane segments, the left and the right
                break;
            }

            if (pixel_value > 0 && !in_lane) {
                // Found a white pixel and we are not inside a lane segment yet.
                // Start of a lane segment then.
                segment_start = x;
                in_lane = true;
            } else if (pixel_value == 0 && in_lane) {
                // transitioning to end of lane segment
                int lane_width = x - segment_start;
                if (lane_width >= config.min_lane_width &&
                    lane_width <= config.max_lane_width) {
                    // End of vail lane segment
                    lane_segments.push_back({segment_start, x - 1});
                    in_lane = false;

                    RCLCPP_DEBUG(get_logger(), "Add Lane Segment that is %d pixels wide",
                                 lane_width);
                } else {
                    // end of the object that is not a lane
                    in_lane = false;
                }
            }
        }

        // Handle case where lane extends to the edge of the image
        if (in_lane && segment_start >= 0) {
            int lane_width = bin_mask.cols - segment_start;
            if (lane_width >= config.min_lane_width &&
                lane_width <= config.max_lane_width) {
                lane_segments.push_back({segment_start, bin_mask.cols - 1});
                RCLCPP_DEBUG(get_logger(), "Add Lane Segment that is %d pixels wide",
                             lane_width);
            }
        }

        // Process lane segments
        if (lane_segments.size() >= 2) {
            // Fix variable assignments: get right edge of left lane and left edge of
            // right lane
            int left_lane_right_edge =
                lane_segments[0].second;  // Right edge of left lane
            int right_lane_left_edge =
                lane_segments.back().first;  // Left edge of right lane
            int road_width = abs(right_lane_left_edge - left_lane_right_edge);
            RCLCPP_DEBUG(get_logger(), "Road width: %d", road_width);

            // check if the road width makes sense
            if (road_width >= config.min_road_width &&
                road_width <= config.max_road_width) {
                // Calculate center between the inner edges of the two lanes

                //  find the mid point between the left and right lane.
                int center_x = (left_lane_right_edge + right_lane_left_edge) / 2;
                this->center_lane.push_back(Point(center_x, y));

                // save left and right points
                this->right_lane.push_back(Point(right_lane_left_edge, y));
                this->left_lane.push_back(Point(left_lane_right_edge, y));

                RCLCPP_DEBUG(
                    get_logger(), "Left, center, right: (%d,%d) (%d,%d) (%d,%d)",
                    left_lane_right_edge, y, right_lane_left_edge, y, center_x, y);
            }
        }

        // do a median filter to remove outlier points
        median_filter();
    }

    // check if we found enough points
    if (this->left_lane.size() > 3) {
        // we found
        return true;
    } else {
        // we found nothing
        return false;
    }
}

// ============== Private Member Function Implementations =====================
void Ld_scanner::median_filter() {
    vector<Point> filtered_center;
    vector<Point> filtered_right;
    vector<Point> filtered_left;

    // get configuration settings for median filtering
    int window_size = config.median_window;
    double threshold = config.median_threshold;

    // go through ever element
    for (size_t i = 0; i < this->center_lane.size(); i++) {
        // calculate current window start and end index
        int start = max(0, (int)i - window_size / 2);
        int end = min((int)center_lane.size() - 1, (int)i + window_size / 2);

        vector<int> window_values;
        for (int j = start; j <= end; j++) {
            if ((size_t)j != i) {
                window_values.push_back(center_lane[j].x);
            }
        }

        // calculated median
        if (!window_values.empty()) {
            auto beginning_iter = window_values.begin();
            auto ending_iter = window_values.end();

            // sort the values in asending order
            sort(beginning_iter, ending_iter);

            // now that is sorted, the middle index is going to be the
            // median
            int median_val = window_values[window_values.size() / 2];

            if (abs(center_lane[i].x - median_val) <= threshold) {
                // check if the current index is within the threshold
                filtered_center.push_back(center_lane[i]);
                filtered_left.push_back(this->left_lane[i]);
                filtered_right.push_back(this->right_lane[i]);
            }
        } else {
            // Keep first/last points if no neighbors
            filtered_center.push_back(center_lane[i]);
            filtered_left.push_back(this->left_lane[i]);
            filtered_right.push_back(this->right_lane[i]);
        }
    }

    if (filtered_center.size() > 0) {
        // assign the left, right, center lanes new filter values
        this->left_lane = filtered_left;
        this->right_lane = filtered_right;
        this->center_lane = filtered_center;
    } else {
        RCLCPP_DEBUG(get_logger(), "The scanner found not vail lanes");
    }
}

}  // namespace ld_scanner_ns

File: src/lane_detection_py/lane_detection_py/ld_scanner.py
import rclpy
import numpy as np
from .ld_config import LdConfig # Assumes ld_config is available in the same folder

class LdScanner:
    def __init__(self, config: LdConfig, logger=None):
        self.config = config
        
        # Lane Data Storage - storing as (x, y) tuples
        self.left_lane = []
        self.right_lane = []
        self.center_lane = []
        
        # NEW: Memory variable for Single-Line Fallback
        # We start with a default guess, but it will dynamically update 
        # the moment it sees two valid lines!
        self.known_lane_width = 350 
        
        # ROS2 Logger
        self.logger = logger or rclpy.logging.get_logger('Ld_scanner')

    def run(self, bin_mask: np.ndarray) -> bool:
        """
        Scans binary images for lane line segments based on width and position.
        Returns True if more than 3 lane points are found.
        """
        # Reset the lane pixel coordinate vectors
        self.left_lane.clear()
        self.right_lane.clear()
        self.center_lane.clear()

        rows, cols = bin_mask.shape

        # Scan rows based on the configured step size
        for y in range(0, rows, self.config.steps):
            lane_segments = []
            
            # Extract the whole row and find indices of all white pixels (>0)
            row = bin_mask[y, :]
            nonzero_indices = np.nonzero(row)[0]
            
            if len(nonzero_indices) > 0:
                # Find gaps greater than 1 to separate contiguous line segments
                split_points = np.where(np.diff(nonzero_indices) > 1)[0] + 1
                segments = np.split(nonzero_indices, split_points)

                for seg in segments:
                    if len(lane_segments) == 2:
                        break # We only need the left and right lane
                    
                    segment_start = seg[0]
                    segment_end = seg[-1]
                    lane_width = segment_end - segment_start
                    
                    # Check if the segment width matches what a lane should look like
                    if self.config.min_lane_width <= lane_width <= self.config.max_lane_width:
                        lane_segments.append((segment_start, segment_end))
            
            if len(lane_segments) >= 2:
                left_lane_right_edge = lane_segments[0][1]
                right_lane_left_edge = lane_segments[-1][0]
                
                road_width = abs(right_lane_left_edge - left_lane_right_edge)
                
                # Check if the road width makes sense
                if self.config.min_road_width <= road_width <= self.config.max_road_width:
                    
                    # MEMORY UPDATE: We see both lines, so memorize this exact width!
                    self.known_lane_width = road_width
                    
                    # Find the midpoint
                    center_x = int((left_lane_right_edge + right_lane_left_edge) / 2)
                    
                    # Store the points
                    self.center_lane.append((center_x, y))
                    self.right_lane.append((right_lane_left_edge, y))
                    self.left_lane.append((left_lane_right_edge, y))
                    
            elif len(lane_segments) == 1:
                seg = lane_segments[0]
                
                # Check if this single line is on the left half or right half of the camera
                if seg[1] < (cols / 2):
                    # It's the LEFT line (Yellow). We lost the right line.
                    left_lane_right_edge = seg[1]
                    # Hallucinate the right line using our memory!
                    right_lane_left_edge = int(left_lane_right_edge + self.known_lane_width)
                else:
                    # It's the RIGHT line (White). We lost the left line.
                    right_lane_left_edge = seg[0]
                    # Hallucinate the left line using our memory!
                    left_lane_right_edge = int(right_lane_left_edge - self.known_lane_width)
                
                # Calculate the center based on our hallucinated line
                center_x = int((left_lane_right_edge + right_lane_left_edge) / 2)
                
                # Store the points just like normal!
                self.center_lane.append((center_x, y))
                self.right_lane.append((right_lane_left_edge, y))
                self.left_lane.append((left_lane_right_edge, y))

        # Do a median filter to remove outlier points 
        self.median_filter()

        # Return True if we found enough points
        return len(self.left_lane) > 3

    def median_filter(self):
        """
        Does a sliding window median filter on the left, right, and center lane scan results.
        """
        if not self.center_lane:
            return

        filtered_center = []
        filtered_left = []
        filtered_right = []
        
        window_size = self.config.median_window
        threshold = self.config.median_threshold
        
        n = len(self.center_lane)
        
        # Go through every element
        for i in range(n):
            # Calculate current window start and end index
            start = max(0, i - window_size // 2)
            end = min(n - 1, i + window_size // 2)
            
            # Get x-values in the window, skipping the current element 'i'
            window_values = [self.center_lane[j][0] for j in range(start, end + 1) if j != i]
            
            if window_values:
                # Calculate median
                median_val = np.median(window_values)
                
                # Check if the current index is within the threshold
                if abs(self.center_lane[i][0] - median_val) <= threshold:
                    filtered_center.append(self.center_lane[i])
                    filtered_left.append(self.left_lane[i])
                    filtered_right.append(self.right_lane[i])
            else:
                # Keep first/last points if no neighbors
                filtered_center.append(self.center_lane[i])
                filtered_left.append(self.left_lane[i])
                filtered_right.append(self.right_lane[i])

        if filtered_center:
            # Assign the new filtered values
            self.left_lane = filtered_left
            self.right_lane = filtered_right
            self.center_lane = filtered_center
        else:
            self.logger.debug("The scanner found no valid lanes after filtering")

    # Getters
    def get_left_lane(self): return self.left_lane
    def get_right_lane(self): return self.right_lane
    def get_center_lane(self): return self.center_lane